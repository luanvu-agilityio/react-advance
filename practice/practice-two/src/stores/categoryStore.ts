import productApi from '@services/product'
import { create } from 'zustand'
import { devtools, subscribeWithSelector } from 'zustand/middleware'

interface CategoryState {
  // Pagination
  currentPage: number
  limit: number
  displayLimit: number

  // Filters
  category?: string
  subcategory?: string
  selectedBrands: string[]
  selectedRatings: number[]
  priceRange: { min: number; max: number }
  searchQuery?: string

  // UI State
  viewMode: string

  //Sorting
  sortBy: string
  sortOrder: 'asc' | 'desc'
  setSortBy: (field: string) => void
  setSortOrder: (order: 'asc' | 'desc') => void

  // Actions
  setPage: (page: number) => void
  setLimit: (limit: number) => void
  setCategory: (category?: string) => void
  setSubcategory: (subcategory?: string) => void
  setBrands: (brands: string[]) => void
  setRatings: (ratings: number[]) => void
  setPriceRange: (range: { min: number; max: number }) => void
  setSearchQuery: (query?: string) => void
  setViewMode: (mode: string) => void
  resetFilters: () => void
  resetPagination: () => void
  setDisplayLimit: (limit: number) => void

  // URL helpers
  getUrlParams: () => URLSearchParams
  setFromUrl: (searchParams: URLSearchParams) => void
  getApiParams: () => Record<string, string | number | string[] | number[]>
  getProductCountBySubcategory: (subcategory: string) => Promise<number>
}

const defaultPriceRange = { min: 0, max: 1000 }

export const useCategoryStore = create<CategoryState>()(
  devtools(
    subscribeWithSelector((set, get) => ({
      // Initial state
      currentPage: 1,
      limit: 5,
      displayLimit: 5,
      selectedBrands: [],
      selectedRatings: [],
      priceRange: defaultPriceRange,
      viewMode: 'grid',
      sortBy: 'name',
      sortOrder: 'asc',

      // Sorting actions
      // Sort actions
      setSortBy: (field: string) => {
        set({ sortBy: field, currentPage: 1 })
      },

      setSortOrder: (order: 'asc' | 'desc') => {
        set({ sortOrder: order, currentPage: 1 })
      },

      // Pagination actions
      setPage: (page: number) => {
        set({ currentPage: Math.max(1, page) })
      },

      setLimit: (limit: number) => {
        set({ limit: Math.max(1, limit), currentPage: 1 })
      },

      // Filter actions
      setCategory: (category) => {
        set({ category, currentPage: 1 }) // Reset page when category changes
      },

      setSubcategory: (subcategory) => {
        set({ subcategory, currentPage: 1 })
      },

      setBrands: (brands) => {
        set({ selectedBrands: brands, currentPage: 1 })
      },

      setRatings: (ratings) => {
        set({ selectedRatings: ratings, currentPage: 1 })
      },

      setPriceRange: (range) => {
        set({ priceRange: range, currentPage: 1 })
      },

      setSearchQuery: (query) => {
        set({ searchQuery: query, currentPage: 1 })
      },

      setViewMode: (mode) => {
        set({ viewMode: mode })
      },

      setDisplayLimit: (displayLimit: number) => {
        const newLimit = Math.max(1, displayLimit)
        set({
          displayLimit: newLimit,
          limit: newLimit,
          currentPage: 1,
        })
      },
      // Reset functions
      resetFilters: () => {
        set({
          selectedBrands: [],
          selectedRatings: [],
          priceRange: defaultPriceRange,
          subcategory: undefined,
          currentPage: 1,
        })
      },

      resetPagination: () => {
        set({ currentPage: 1 })
      },

      // URL management
      getUrlParams: () => {
        const state = get()
        const params = new URLSearchParams()

        if (state.currentPage > 1) {
          params.set('page', state.currentPage.toString())
        }

        // Always include limit in URL params
        params.set('limit', state.limit.toString())

        if (state.subcategory) {
          params.set(
            'subcategory',
            state.subcategory.toLowerCase().replace(/\s+/g, '-')
          )
        }

        if (state.searchQuery) {
          params.set('search', state.searchQuery)
        }

        params.set('sort', state.sortBy)
        params.set('order', state.sortOrder)

        return params
      },

      setFromUrl: (searchParams) => {
        const page = parseInt(searchParams.get('page') || '1', 10)
        const limit = parseInt(searchParams.get('limit') || '5', 10)
        const subcategory = searchParams.get('subcategory')
        const search = searchParams.get('search')
        const sort = searchParams.get('sort') || 'name'
        const order = searchParams.get('order') === 'desc' ? 'desc' : 'asc'

        set({
          currentPage: Math.max(1, page),
          limit: Math.max(1, limit),
          subcategory: subcategory || undefined,
          searchQuery: search || undefined,
          sortBy: sort,
          sortOrder: order as 'asc' | 'desc',
        })
      },

      // API params
      getApiParams: () => {
        const state = get()
        const params: Record<string, string | number | string[] | number[]> = {
          p: state.currentPage,
          l: state.limit,
        }
        if (state.category) params.category = state.category
        if (state.subcategory) params.subcategory = state.subcategory
        if (state.selectedBrands.length > 0)
          params.brands = state.selectedBrands
        if (state.priceRange.min > defaultPriceRange.min)
          params.minPrice = state.priceRange.min
        if (state.priceRange.max < defaultPriceRange.max)
          params.maxPrice = state.priceRange.max
        if (state.searchQuery) params.search = state.searchQuery
        if (state.selectedRatings.length > 0)
          params.ratings = state.selectedRatings

        params.sortBy = state.sortBy
        params.sortOrder = state.sortOrder
        return params
      },

      getProductCountBySubcategory: async (subcategory: string) => {
        try {
          // Make a specific API call to count products
          const response = await productApi.getProductCount({ subcategory })
          return response
        } catch (error) {
          console.error('Error getting product count:', error)
          return 0
        }
      },
    })),
    { name: 'category-store' }
  )
)
