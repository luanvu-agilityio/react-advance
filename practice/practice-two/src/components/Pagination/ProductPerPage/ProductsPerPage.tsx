'use client'

import { useCategoryStore } from '@stores/categoryStore'
import { SelectWrapper } from '@components/Searchbar/Searchbar.styles'
import Select from '@components/common/Select'
import { FilterControlContainer, FilterLabel } from '../PaginationStyles'

const productsPerPageOptions = [
  { value: '5', label: '5 per page' },
  { value: '10', label: '10 per page' },
  { value: '15', label: '15 per page' },
  { value: '20', label: '20 per page' },
]

export const ProductsPerPage = () => {
  const { displayLimit, setDisplayLimit } = useCategoryStore()

  const handleChange = (value: string) => {
    const newLimit = parseInt(value, 10)
    setDisplayLimit(newLimit)
  }

  return (
    <FilterControlContainer>
      <FilterLabel>Product per page</FilterLabel>
      <SelectWrapper
        style={{
          width: '100%',
          margin: '0',
          padding: 0,
          minWidth: '120px',
          maxWidth: '200px',
        }}
      >
        <Select
          options={productsPerPageOptions}
          value={displayLimit.toString()}
          onChange={handleChange}
          variant="filter"
        />
      </SelectWrapper>
    </FilterControlContainer>
  )
}
