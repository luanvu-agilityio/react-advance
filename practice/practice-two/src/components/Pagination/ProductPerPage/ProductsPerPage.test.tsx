import { fireEvent, render, screen } from '@testing-library/react'
import { ProductsPerPage } from './ProductsPerPage'
import { useCategoryStore } from '@stores/categoryStore'

// Mock the Zustand store
jest.mock('@stores/categoryStore', () => ({
  useCategoryStore: jest.fn(),
}))

// Mock the Select component
jest.mock('@components/common/Select', () => {
  return {
    __esModule: true,
    default: ({
      value,
      onChange,
      options,
    }: {
      value: string
      onChange: (value: string) => void
      options: Array<{ value: string; label: string; disabled?: boolean }>
    }) => (
      <div data-testid="mock-select">
        <span data-testid="current-value">{value}</span>
        {options.map((option) => (
          <button
            key={option.value}
            data-testid={`option-${option.value}`}
            onClick={() => onChange(option.value)}
            disabled={option.disabled}
          >
            {option.label}
          </button>
        ))}
      </div>
    ),
  }
})

describe('ProductsPerPage', () => {
  const mockSetDisplayLimit = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()
    // Default mock implementation
    ;(useCategoryStore as unknown as jest.Mock).mockReturnValue({
      displayLimit: 5,
      setDisplayLimit: mockSetDisplayLimit,
    })
  })

  // SNAPSHOT TESTS FOR RENDERING
  describe('Rendering', () => {
    it('renders different component states correctly', () => {
      // Default state (5 per page)
      const { container: defaultContainer } = render(<ProductsPerPage />)
      expect(defaultContainer).toMatchSnapshot('default-5-per-page')

      // With 10 per page selected
      ;(useCategoryStore as unknown as jest.Mock).mockReturnValue({
        displayLimit: 10,
        setDisplayLimit: mockSetDisplayLimit,
      })
      const { container: with10PerPage } = render(<ProductsPerPage />)
      expect(with10PerPage).toMatchSnapshot('with-10-per-page')

      // With 15 per page selected
      ;(useCategoryStore as unknown as jest.Mock).mockReturnValue({
        displayLimit: 15,
        setDisplayLimit: mockSetDisplayLimit,
      })
      const { container: with15PerPage } = render(<ProductsPerPage />)
      expect(with15PerPage).toMatchSnapshot('with-15-per-page')

      // With 20 per page selected
      ;(useCategoryStore as unknown as jest.Mock).mockReturnValue({
        displayLimit: 20,
        setDisplayLimit: mockSetDisplayLimit,
      })
      const { container: with20PerPage } = render(<ProductsPerPage />)
      expect(with20PerPage).toMatchSnapshot('with-20-per-page')

      // With invalid value
      ;(useCategoryStore as unknown as jest.Mock).mockReturnValue({
        displayLimit: NaN,
        setDisplayLimit: mockSetDisplayLimit,
      })
      const { container: withInvalidValue } = render(<ProductsPerPage />)
      expect(withInvalidValue).toMatchSnapshot('with-invalid-value')
    })
  })

  // INTERACTIVE BEHAVIOR TESTS
  describe('Interactive behavior', () => {
    it('calls setDisplayLimit with the correct numeric value when an option is selected', () => {
      render(<ProductsPerPage />)

      // Click option for 10 per page
      fireEvent.click(screen.getByTestId('option-10'))
      expect(mockSetDisplayLimit).toHaveBeenCalledWith(10)

      // Click option for 15 per page
      fireEvent.click(screen.getByTestId('option-15'))
      expect(mockSetDisplayLimit).toHaveBeenCalledWith(15)

      // Click option for 20 per page
      fireEvent.click(screen.getByTestId('option-20'))
      expect(mockSetDisplayLimit).toHaveBeenCalledWith(20)
    })

    it('displays the current displayLimit from the store', () => {
      // Set displayLimit to 15 in the store
      ;(useCategoryStore as unknown as jest.Mock).mockReturnValue({
        displayLimit: 15,
        setDisplayLimit: mockSetDisplayLimit,
      })

      render(<ProductsPerPage />)

      // Check that the current value is '15'
      expect(screen.getByTestId('current-value').textContent).toBe('15')
    })
  })
})
