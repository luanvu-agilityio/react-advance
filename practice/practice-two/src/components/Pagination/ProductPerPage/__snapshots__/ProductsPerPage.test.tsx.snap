// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsPerPage Rendering renders different component states correctly: default-5-per-page 1`] = `
<div>
  <div
    class="sc-fpEFIB ksmUmw"
    style="min-width: 180px; height: 60px;"
  >
    <span
      class="sc-gAqISa kDWsLg"
    >
      Product per page
    </span>
    <div
      class="sc-fWnslK jxddlX"
      style="width: auto; min-width: 100px; margin: 0px; padding: 0px;"
    >
      <div
        data-testid="mock-select"
      >
        <span
          data-testid="current-value"
        >
          5
        </span>
        <button
          data-testid="option-5"
        >
          5 per page
        </button>
        <button
          data-testid="option-10"
        >
          10 per page
        </button>
        <button
          data-testid="option-15"
        >
          15 per page
        </button>
        <button
          data-testid="option-20"
        >
          20 per page
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ProductsPerPage Rendering renders different component states correctly: with-10-per-page 1`] = `
<div>
  <div
    class="sc-fpEFIB ksmUmw"
    style="min-width: 180px; height: 60px;"
  >
    <span
      class="sc-gAqISa kDWsLg"
    >
      Product per page
    </span>
    <div
      class="sc-fWnslK jxddlX"
      style="width: auto; min-width: 100px; margin: 0px; padding: 0px;"
    >
      <div
        data-testid="mock-select"
      >
        <span
          data-testid="current-value"
        >
          10
        </span>
        <button
          data-testid="option-5"
        >
          5 per page
        </button>
        <button
          data-testid="option-10"
        >
          10 per page
        </button>
        <button
          data-testid="option-15"
        >
          15 per page
        </button>
        <button
          data-testid="option-20"
        >
          20 per page
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ProductsPerPage Rendering renders different component states correctly: with-15-per-page 1`] = `
<div>
  <div
    class="sc-fpEFIB ksmUmw"
    style="min-width: 180px; height: 60px;"
  >
    <span
      class="sc-gAqISa kDWsLg"
    >
      Product per page
    </span>
    <div
      class="sc-fWnslK jxddlX"
      style="width: auto; min-width: 100px; margin: 0px; padding: 0px;"
    >
      <div
        data-testid="mock-select"
      >
        <span
          data-testid="current-value"
        >
          15
        </span>
        <button
          data-testid="option-5"
        >
          5 per page
        </button>
        <button
          data-testid="option-10"
        >
          10 per page
        </button>
        <button
          data-testid="option-15"
        >
          15 per page
        </button>
        <button
          data-testid="option-20"
        >
          20 per page
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ProductsPerPage Rendering renders different component states correctly: with-20-per-page 1`] = `
<div>
  <div
    class="sc-fpEFIB ksmUmw"
    style="min-width: 180px; height: 60px;"
  >
    <span
      class="sc-gAqISa kDWsLg"
    >
      Product per page
    </span>
    <div
      class="sc-fWnslK jxddlX"
      style="width: auto; min-width: 100px; margin: 0px; padding: 0px;"
    >
      <div
        data-testid="mock-select"
      >
        <span
          data-testid="current-value"
        >
          20
        </span>
        <button
          data-testid="option-5"
        >
          5 per page
        </button>
        <button
          data-testid="option-10"
        >
          10 per page
        </button>
        <button
          data-testid="option-15"
        >
          15 per page
        </button>
        <button
          data-testid="option-20"
        >
          20 per page
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ProductsPerPage Rendering renders different component states correctly: with-invalid-value 1`] = `
<div>
  <div
    class="sc-fpEFIB ksmUmw"
    style="min-width: 180px; height: 60px;"
  >
    <span
      class="sc-gAqISa kDWsLg"
    >
      Product per page
    </span>
    <div
      class="sc-fWnslK jxddlX"
      style="width: auto; min-width: 100px; margin: 0px; padding: 0px;"
    >
      <div
        data-testid="mock-select"
      >
        <span
          data-testid="current-value"
        >
          NaN
        </span>
        <button
          data-testid="option-5"
        >
          5 per page
        </button>
        <button
          data-testid="option-10"
        >
          10 per page
        </button>
        <button
          data-testid="option-15"
        >
          15 per page
        </button>
        <button
          data-testid="option-20"
        >
          20 per page
        </button>
      </div>
    </div>
  </div>
</div>
`;
